README.rst
|Build Status| |Cover alls| |Gemnasium|


Proyectos de Ley del Congreso
=============================

`http://proyectosdeley.pe` es un intento de transparentar el Congreso y poner
al alcance de la mayor cantidad de personas los proyectos de ley presentados y
discutidos en el parlamento. La información mostrada es tomada directamente de
la página web del Congreso.

Esta página ha sido desarrollada en forma independiente por la ONG Hiperderecho
y la asociación Contribuyentes por Respeto, organizaciones sin fines de lucro
dedicadas a investigar, facilitar el entendimiento público y promover
el respeto de los derechos y libertades en entornos digitales.

`http://proyectosdeley.pe`

Configuración
-------------

Esta aplicación se basa en el siguiente *software*:

* Python v3.4
* django v1.8.4
* PostgreSQL
* elasticsearch

Crear y un activar un virtualenv_ que use Python3. En un terminal, instalar 
las dependencias para desarrollo local::

    $ pip install -r requirements/dev.txt

.. _virtualenv: http://docs.python-guide.org/en/latest/dev/virtualenvs/

Para correr el software, reemplazar ``yourapp`` con el nombre
del folder que contiene el proyecto de Django::

    $ python yourapp/manage.py runserver --settings=yourapp.settings.local

Recasting DateField as DateTimeField::

    alter table pdl_proyecto alter column fecha_presentacion TYPE timestamp using fecha_presentacion::timestamp;

Obteniendo información
----------------------
Toda la info se obtiene desde las páginas web del Congreso mediante el uso de
un *scraper* basado en el *framework* Scrapy. El *scraper* se encuentra en
este repositorio_.

.. _repositorio: https://github.com/proyectosdeley/proyectos_de_ley_scraper

Custom commands
---------------

Para generar una tabla resúmen del número de proyectos que ha quedado estancado
en cada comisión del Congreso::

    > python manage.py create_stats --settings=proyectos_de_ley.settings.local

Plantilla HTML
--------------
* El archivo ``base.html`` funciona como plantilla para crear las páginas HTML.
  Cualquier cambio al estilo se debe realizar en este archivo. Esta plantilla
  usa un estilo basado en Twitter Bootstrap con *responsive features* para que
  se vea bien en computadoras y dispositivos móbiles.
* Esos campos se usan para introducir en contenido en la plantilla y generar
  los archivos HTML.

.. |Build Status| image:: https://travis-ci.org/proyectosdeley/proyectos_de_ley.svg?branch=master
   :target: https://travis-ci.org/proyectosdeley/proyectos_de_ley
.. |Cover alls| image:: https://coveralls.io/repos/proyectosdeley/proyectos_de_ley/badge.png
   :target: https://coveralls.io/r/proyectosdeley/proyectos_de_ley
.. |Gemnasium| image:: https://gemnasium.com/proyectosdeley/proyectos_de_ley.svg
   :target: https://gemnasium.com/proyectosdeley/proyectos_de_ley



HISTORY.rst
History
=======

v2.2.0 (2016-09-07)
~~~~~~~~~~~~~~~~~~~
* agregada capacidad de soportar proyectos de Legislatura 2011 y 2016.
* agregada interface administrador en Django.
* mostrar proyectos de actual legislatura y menor prominencia a proyectos de
  legislaturas pasadas.
* actualización de spiders.
* actualización de dependencias: Django, rest_framework, rest_framework_swagger.
* actualización de layout de docs en API.
* bug fixed: número de proyectos aprovados en front-page
* bug fixed: búsqueda de proyecto "209" resultará en búsqueda de proyecto 00209-2016.
* bug fixed: ya no se muestra código 404 en el API.
* bug fixed: logo de Hiperderecho está hosteado en nuestro server.
* bug fixed: lista de proyectos de legislatura pasada ya no genera error 500.
* bug fixed: modificación de modelos para casos en que congresistas tengan nombres
  con y sin tildes. Cualquier variación mostrará proyectos de cada congresista.

v2.1.1 (2015-10-29)
~~~~~~~~~~~~~~~~~~~
* script para hacer backups de base de datos
* Stats. Agregado "sin votación en el pleno" para velocímetro "proyectos con
  dictamen pero sin votación". Además agregué explicación para ese número.
* Búsqueda avanzada por comisiones dan el mismo número de proyectos que los que
  aparecen en el cuadro de estadísticas.
* Arreglado bug en paginación de búsqueda avanzada. Gracias a @skraelings.

v2.1.0 (2015-07-16)
~~~~~~~~~~~~~~~~~~~
* Footer: ahora aparece en posición inferior en todas las páginas.
* Stats: "proyectos con dictamen pero sin votación"
* Stats: reemplazados los 2 primeros velocímetros.
* Stats: ahora muestra exonerados de dictamen.
* Stats: corregidos números de proyetos exonerados de 2da votación.
* Stats: corregido número de proyectos que aún no son ley.
* Búsqueda avanzada: permite buscar combinación de campos como: fechas, palabra
  clave, comisión, nombre de congresista y grupo parlamentario.
* Corregido: bug cuando query contenía espacios en blanco el navegador entraba
  en loop infinito.
* Corregido: número de proyectos que no son ley (considerando que en las leyes
  promulgadas hay numerosos proyectos que han sido fusionados).
* API: tenemos API que emite datos en formato abierto en JSON y CSV.
  Ver http://proyectosdeley.pe/docs/
* API: documentación automática.
* API: nuevo ícono para acceder a la documentación en la barra de navegación.
* Corregido schema de la base de datos.
* Nueva columna `ascii` en tabla Slug.

v2.0.1 (2015-03-31)
~~~~~~~~~~~~~~~~~~~
* Fixed with of stats counter.
* Better search engine. Now query words will match accented and unaccented words.
* Stats: "en comisión sin dictamen"
* Stats: "sin tramitar"

v2.0.0 (2015-03-30)
~~~~~~~~~~~~~~~~~~~
* Using haystack and elasticsearch.
* Indexing ascii and accented chars for search fields by overriding
  the ``prepare_text`` function in Haystack.
* Stats: chart "by commission" has links on the bars. Click them to see projects
  stuck in each commission'
* Advanced search: now we can search projects by comission.
* Update stats.

v1.7.0 (2015-02-04)
~~~~~~~~~~~~~~~~~~~
* Humanized dates using ``arrow``.
* Added *legislatura*.

v1.6.0 (2015-01-31)
~~~~~~~~~~~~~~~~~~~
* Fixed bug that showed number of authors + 1 in `seguimientos`.
* Updated about page.
* Timeline starts at end.
* Pull proyectosdeley logo from morachimo domain.
* Show stats in frontpage.
* Fix title of stats page.
* Added counter for number of authors of project in frontpage and project page.

v1.5.3 (2014-12-28)
~~~~~~~~~~~~~~~~~~~
Stats: fixing bug for "otros proyectos dispensados".

v1.5.2 (2014-12-24)
~~~~~~~~~~~~~~~~~~~
Adding respeto.pe logo, new proyectosdeley log and update "About" page.

v1.5.1 (2014-12-16)
~~~~~~~~~~~~~~~~~~~
Fixing create_stats custom command. The update_or_create
command was creating an extra set of items in the database instead of doing
an update. Fixed.

v1.5.0 (2014-12-11)
~~~~~~~~~~~~~~~~~~~
* Only accept as search keywords strings with len > 2.
* Sort simple search result by codigo desc.
* Fix bug on pagination links.
* Keywords were shown with brackets.
* Search box of simple search show user's query.
* Added links to PDF and Expediente in Seguimientos page.
* Added custom command in `stats` app: `create_stats` which should run once a day.
* Added charts to stats page: Number of projects in comisiones and Number of projects that
  did not go to 2nd round of votes. New table expedients to keep events and
  URLs from the `Expediente` page.

v1.4.1 (2014-11-11)
~~~~~~~~~~~~~~~~~~~
* Improved general search engine, also events in `seguimientos` are queried.
* Autofocus on search box when page loads.
* Better highlighting of keywords.
* Check errors in datefield widget (advanced search).
* Show error fields in datefield widget.
* Resize advanced search according to screen sizes.
* Add link to production site.

v1.4.0 (2014-11-08)
~~~~~~~~~~~~~~~~~~~
* Advanced search by *fecha presentación*.
* Stats.
* Improved RSS feed.
* Now we have our own `Seguimiento` page.
* Many more additional fields are scrapped (*seguimientos*, *iniciativas agrupadas*).
* Better scrapping of PDF urls.
* The scrapping functions have been moved to another project: `proyectos_de_ley_scraper`.
* Added version to footer of pages.

v1.2.0 (2014-09-24)
~~~~~~~~~~~~~~~~~~~
* Scrapping more metadata from seguimiento_page.
* Killed bug to get PDF url when the filename includes funny characters.
* Custom command to update `seguimiento` events for each project in our database.

v1.1.1 (2014-09-22)
~~~~~~~~~~~~~~~~~~~
* Favicon.
* Don't show navigation bar if there are no results.

v1.1.0 (2014-09-21)
~~~~~~~~~~~~~~~~~~~
Pagination for search results (40 items per page).

v1.0.1 (2014-09-20)
~~~~~~~~~~~~~~~~~~~
Own pagination for index and congresista pages.

v1.0.0 (2014-09-15)
~~~~~~~~~~~~~~~~~~~
Migrated to Django.


proyectos_de_ley/pdl/utils.py
import datetime
import re
import time
import unicodedata

import arrow

from django.conf import settings
from django.core.paginator import Paginator
from django.core.paginator import EmptyPage
from django.core.paginator import PageNotAnInteger

from pdl.models import Proyecto
from pdl.models import Slug


class Timer(object):
    def __init__(self, verbose=False):
        self.verbose = verbose

    def __enter__(self):
        self.start = time.time()
        return self

    def __exit__(self, *args):
        self.end = time.time()
        self.secs = self.end - self.start
        self.msecs = self.secs * 1000  # millisecs
        if self.verbose:
            print('elapsed time: %f ms' % self.msecs)


def convert_date_to_string(fecha):
    try:
        nueva_fecha = datetime.date.strftime(fecha, '%m/%d/%Y')
        return nueva_fecha
    except TypeError:
        return None


def convert_string_to_time(string):
    if isinstance(string, str):
        this_time = re.sub("\+[0-9]+$", "", string)
        try:
            time_object = datetime.datetime.strptime(this_time, "%Y-%m-%d")
            return time_object
        except ValueError:
            pass

        try:
            time_object = datetime.datetime.strptime(this_time, "%Y-%m-%d %H:%M:%S.%f")
        except TypeError:
            # This exception is only for our test that wants str not date obj
            time_object = item.time_created
        except ValueError:
            time_object = datetime.datetime.strptime(this_time, "%Y-%m-%d %H:%M:%S")

        return time_object
    else:
        # is should be a date object
        return string


def prettify_item(item):
    fecha_pres_humana = arrow.get(item.fecha_presentacion).format('DD MMMM, YYYY', locale='es_es')
    out = "<p>"
    out += "<a href='/p/" + str(item.short_url)
    out += "' title='Permalink'>"
    out += "<b>" + item.numero_proyecto + "</b></a> "
    out += "&nbsp;"
    out += "<small>Presentado: " + fecha_pres_humana + "</small></p>\n"
    out += "<h4>" + item.titulo + "</h4>\n"

    if len(item.congresistas) > 0:
        out += "Autores <span class='badge'>" + str(len(item.congresistas.split(";"))) + "</span>\n"
    out += "<p>" + hiperlink_congre(item.congresistas) + "</p>\n"

    if item.pdf_url != '':
        out += "<a class='btn btn-lg btn-primary'"
        out += " href='" + str(item.pdf_url) + "' role='button'>PDF</a>\n"
    else:
        out += "<a class='btn btn-lg btn-primary disabled'"
        out += " href='#' role='button'>Sin PDF</a>\n"

    if item.expediente != '':
        out += "<a class='btn btn-lg btn-primary'"
        out += " href='" + item.expediente
        out += "' role='button'>EXPEDIENTE</a>\n"
    else:
        out += "<a class='btn btn-lg btn-primary disabled'"
        out += " href='#' role='button'>Sin EXPEDIENTE</a>\n"

    if item.seguimiento_page != '':
        out += "<a class='btn btn-lg btn-primary'"
        out += " href='/p/" + item.short_url + "/seguimiento"
        out += "' role='button'>Seguimiento</a>"
    return out


def hiperlink_congre(congresistas):
    # tries to make a hiperlink for each congresista name to its own webpage
    for name in congresistas.split("; "):
        link = "<a href='/congresista/"
        link += str(convert_name_to_slug(name))
        link += "' title='ver todos sus proyectos'>"
        link += name + "</a>"
        congresistas = congresistas.replace(name, link)
    congresistas = congresistas.replace("; ", ";\n")
    return congresistas

def convert_name_to_slug(name):
    """Takes a congresista name and returns its slug."""
    name = name.strip()
    name = name.replace(",", "").lower()
    name = name.split(" ")

    if len(name) > 2:
        i = 0
        slug = ""
        while i < 3:
            slug += name[i]
            if i < 2:
                slug += "_"
            i += 1
        slug = unicodedata.normalize('NFKD', slug).encode('ascii', 'ignore')
        slug = str(slug, encoding="utf-8")
        return slug + "/"


def do_pagination(request, all_items, search=False, advanced_search=None):
    """
    :param request: contains the current page requested by user
    :param all_items:
    :param search: if search is False items will be prettified in long form.
           if search is True then items will be prettified as small items
           for search results.
    :param advanced_search: True or None to point out that we come from the
           advanced search page.
    :return: dict containing paginated items and pagination bar
    """
    if 'comision' in request.GET and request.GET['comision'] != '':
        comision = request.GET['comision']
    else:
        comision = False

    if search is False:
        paginator = Paginator(all_items, 20)
    else:
        paginator = Paginator(all_items, 40)

    page = request.GET.get('page')

    try:
        items = paginator.page(page)
        cur = int(page)
    except PageNotAnInteger:
        # If page is not an integer, deliver first page.
        items = paginator.page(1)
        cur = 1
    except EmptyPage:
        # If page is out of range (e.g. 9999), deliver last page of results.
        items = paginator.page(paginator.num_pages)
        cur = 1

    pretty_items = []
    for i in items.object_list:
        if search is False:
            pretty_items.append(prettify_item(i))
        else:
            if settings.TESTING:
                pretty_items.append(prettify_item_small(i))
            elif advanced_search is True:
                pretty_items.append(prettify_item_small(i))
            else:
                pretty_items.append(prettify_item_small(i.object))

    if cur > 20:
        first_half = range(cur - 10, cur)
        # is current less than last page?
        if cur < paginator.page_range[-1] - 10:
            second_half = range(cur + 1, cur + 10)
        else:
            second_half = range(cur + 1, paginator.page_range[-1])
    else:
        first_half = range(1, cur)
        if paginator.page_range[-1] > 20:
            second_half = range(cur + 1, 21)
        else:
            second_half = range(cur + 1, paginator.page_range[-1] + 1)

    obj = {
        'items': items,
        "pretty_items": pretty_items,
        "first_half": first_half,
        "second_half": second_half,
        "first_page": paginator.page_range[0],
        "last_page": paginator.page_range[-1],
        "current": cur,
        "comision": comision,
    }
    return obj


def find_slug_in_db(congresista_slug):
    try:
        item = Slug.objects.get(slug=congresista_slug)
        return item.ascii
    except Slug.DoesNotExist:
        try:
            congresista_slug += '/'
            item = Slug.objects.get(slug=congresista_slug)
            return item.ascii
        except Slug.DoesNotExist:
            return None


def get_last_items():
    """All items from the database are extracted as list of dictionaries."""
    items = Proyecto.objects.all().order_by('-codigo')[:20]
    pretty_items = []
    for i in items:
        pretty_items.append(prettify_item(i))
    return pretty_items


def prettify_item_small(item):
    out = "<p><a href='/p/" + item.short_url
    out += "' title='Permalink'>"
    out += item.codigo
    out += "</a>\n "
    out += item.titulo
    if item.pdf_url != '' and item.pdf_url is not None:
        out += '\n <span class="glyphicon glyphicon-cloud-download"></span>'
        out += ' <a href="' + item.pdf_url + '">PDF</a>'
    else:
        out += ' [sin PDF]'

    if item.expediente != '' and item.expediente is not None:
        out += '\n <span class="glyphicon glyphicon-link"></span>'
        out += ' <a href="' + item.expediente + '">Expediente</a>'
    else:
        out += ' [sin Expediente]'

    if item.seguimiento_page != '' and item.seguimiento_page is not None:
        out += '\n <span class="glyphicon glyphicon-link"></span>'
        out += ' <a href="/p/' + item.short_url + \
               '/seguimiento">Seguimiento</a>'
    else:
        out += '\n [sin Seguimiento]'
    out += '</p>'
    return out
